packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -99,44 +125,117 @@ import {StrategyMath} from "./base/StrategyMath.sol";
       * @notice the owner batch migrates all the crab v1 shares in this contract to crab v2 and initializes 
       * the v2 contract at the same collateral ratio as the v1 contract. 
       */
-     function batchMigrate() external onlyOwner beforeMigration { 
+     function batchMigrate() external onlyOwner beforeMigration {
         // 1. update isMigrated
         isMigrated = true;
 
-         // 2. flash floan eth from euler eq to amt 
+        // 2. flash floan eth from euler eq to amt
+        uint256 crabV1Balance = crabV1.balanceOf(address(this));
+        uint256 crabV1Supply = crabV1.totalSupply();
+        (address _, uint256 id, uint256 totalCollateral, uint256 totalShort) = crabV1.getVaultDetails();
+        uint256 amountEthToBorrow = totalCollateral.wmul(crabV1Balance.wdiv(crabV1Supply));
         bytes memory data;
-        euler.deferLiquidityCheck(address(this), data);
+        euler.deferLiquidityCheck(
+            address(this),
+            abi.encode(
+                FlashloanCallbackData({
+                    caller: msg.sender,
+                    amountToBorrow: amountEthToBorrow,
+                    callSource: uint8(FLASH_SOURCE.BATCH_MIGRATE),
+                    callData: data
+                })
+            )
+        );
 
         // 3. record totalV2Shares
         totalCrabV2SharesReceived = crabV2.balanceOf(address(this));
+    }
 
-     }
+     function onDeferredLiquidityCheck(bytes memory _encodedData) external {
+        require(msg.sender == EULER_MAINNET, "M4");
 
-     function onDeferredLiquidityCheck(bytes memory encodedData) external {
-        require(msg.sender == EULER_MAINNET, "M4"); 
+        FlashloanCallbackData memory data = abi.decode(_encodedData, (FlashloanCallbackData));
 
         // 1. Borrow weth
-        uint256 crabV1Balance = crabV1.balanceOf(address(this));
-        uint256 crabV1Supply = crabV1.totalSupply();
-        (address _, uint256 id, uint256 totalCollateral, uint256 totalShort) = crabV1.getVaultDetails();
-        uint256 amountEthToBorrow = totalCollateral.wmul(crabV1Balance.wdiv(crabV1Supply));
+        IEulerDToken(dToken).borrow(0, data.amountToBorrow);
+        weth.withdraw(data.amountToBorrow);
+
+        // 2. Callback
+        _flashCallback(data.caller, data.amountToBorrow, data.callSource, data.callData);
+
+        // 3. Repay the weth:
+        weth.deposit{value: data.amountToBorrow}();
+        weth.approve(EULER_MAINNET, type(uint256).max);
+        IEulerDToken(dToken).repay(0, data.amountToBorrow);
+    }
+
+    function _flashCallback(
```

Comment:
why do we have an extra callback? Is this getting called back from anywhere else or just the from onDeferredLiquidityCheck?

Reply:
It's called from onDeferredLiquidityCheck, Just to process the callback. Writing everything in same function looks hard to read, so split it into 2 functions. onDeferredLiquidityCheck will handle Euler things and _flashCallback will process the callback

Reply:
ah ok, Would it make sense to just inherit EulerControllerHelper like ControllerHelper does? Then just have _flashLoan() and _flashCallback()?

Reply:
nvm it works fine